oci/client/push.go

```diff
@@ -34,36 +34,101 @@ import (
 	"github.com/fluxcd/pkg/oci"
 )
 
+// PushOptions are options for configuring the Push operation.
+type PushOptions struct {
+	skipArchive  bool
+	mediaTypeExt string
+	ignorePaths  []string
+	meta         Metadata
+}
+
+// PushOption is a function for configuring PushOptions.
+type PushOption func(o *PushOptions)
+
+// WithSkipArchive skips archiving the source to a tarball and the file is
+// copied directly to the image layer. When this option is specified, the sourcePath
+// argument must be a file and not a directory.
+func WithSkipArchive() PushOption {
+	return func(o *PushOptions) {
+		o.skipArchive = true
+	}
+}
+
+// WithMediaTypeExtension configures the media type extension for the image layer.
+// The final media type will be prefixed with `application/vnd.cncf.flux.content.v1`
+func WithMediaTypeExtension(extension string) PushOption {
+	return func(o *PushOptions) {
+		o.mediaTypeExt = extension
+	}
+}
+
+// WithIgnorePaths configures ignore paths for PushOptions
+func WithIgnorePaths(paths []string) PushOption {
+	return func(o *PushOptions) {
+		o.ignorePaths = paths
+	}
+}
+
+func WithMetadata(meta Metadata) PushOption {
+	return func(o *PushOptions) {
+		o.meta = meta
+	}
+}
+
 // Push creates an artifact from the given directory, uploads the artifact
 // to the given OCI repository and returns the digest.
-func (c *Client) Push(ctx context.Context, url, sourceDir string, meta Metadata, ignorePaths []string) (string, error) {
+func (c *Client) Push(ctx context.Context, url, sourcePath string, opts ...PushOption) (string, error) {
+	o := &PushOptions{}
+	for _, opt := range opts {
+		opt(o)
+	}
 	ref, err := name.ParseReference(url)
 	if err != nil {
 		return "", fmt.Errorf("invalid URL: %w", err)
 	}
 
-	tmpDir, err := os.MkdirTemp("", "oci")
-	if err != nil {
-		return "", err
+	var layerFile string
+	if !o.skipArchive {
+		tmpDir, err := os.MkdirTemp("", "oci")
+		if err != nil {
+			return "", err
+		}
+		defer os.RemoveAll(tmpDir)
+
+		layerFile = filepath.Join(tmpDir, "artifact.tgz")
+		if err := c.Build(layerFile, sourcePath, o.ignorePaths); err != nil {
+			return "", err
+		}
+	} else {
+		// we only want a single file when we aren't archiving
+		f, err := os.Stat(sourcePath)
+		if err != nil {
+			return "", err
+		}
+		if f.IsDir() {
+			return "", fmt.Errorf("can only specify a single file when not compressing")
+		}
+		layerFile = sourcePath
```

Comment:
I don't think it requires a separate function since all we want to do is check if the path is a directory. It can also be tested with the current setup.